pkg/storage/replica_command.go

```diff
@@ -1418,29 +1427,48 @@ func evalGC(
 		return EvalResult{}, err
 	}
 
-	newThreshold, err := cArgs.EvalCtx.GCThreshold()
-	if err != nil {
-		return EvalResult{}, err
-	}
-	newTxnSpanGCThreshold, err := cArgs.EvalCtx.TxnSpanGCThreshold()
-	if err != nil {
-		return EvalResult{}, err
-	}
 	// Protect against multiple GC requests arriving out of order; we track
 	// the maximum timestamps.
-	newThreshold.Forward(args.Threshold)
-	newTxnSpanGCThreshold.Forward(args.TxnSpanGCThreshold)
+
+	var newThreshold hlc.Timestamp
+	if args.Threshold != (hlc.Timestamp{}) {
+		oldThreshold, err := cArgs.EvalCtx.GCThreshold()
+		if err != nil {
+			return EvalResult{}, err
+		}
+		newThreshold = oldThreshold
+		newThreshold.Forward(args.Threshold)
+	}
+
+	var newTxnSpanGCThreshold hlc.Timestamp
+	if args.TxnSpanGCThreshold != (hlc.Timestamp{}) {
+		oldTxnSpanGCThreshold, err := cArgs.EvalCtx.TxnSpanGCThreshold()
+		if err != nil {
+			return EvalResult{}, err
+		}
+		newTxnSpanGCThreshold = oldTxnSpanGCThreshold
+		newTxnSpanGCThreshold.Forward(args.TxnSpanGCThreshold)
+	}
 
 	var pd EvalResult
-	pd.Replicated.State.GCThreshold = newThreshold
-	pd.Replicated.State.TxnSpanGCThreshold = newTxnSpanGCThreshold
+	stateLoader := makeReplicaStateLoader(cArgs.EvalCtx.RangeID())
```

Comment:
Small cherry-pick conflict here because release-1.0 still uses the `stateLoader()` on EvalCtx.